content/posts/csv-boilerplate-for-go.md

```diff
@@ -0,0 +1,307 @@
+	}
+	// remaining columns ...
+	return nil
+}
+
+func (c *Car) MarshalRecord() ([]string, error) {
+	return []string{
+		c.Car,
+		fmt.Sprintf("%f", c.MPG),
+		fmt.Sprintf("%d", c.Cylinders),
+		// remaining columns ...
+	}, nil
+}
+```
+
+This works, but adding, removing, or reordering any of the columns requires us to modify our code in 4 to 5 different places, which we wanted to avoid. So let's try some abstraction that allows us to declaratively specify our columns and how to marshal/unmarshal them:
+
+```go
+type carColumn struct {
+	Name           string
+	Type           string
+	UnmarshalValue func(*Car, string) error
+	MarshalValue   func(*Car) (string, error)
+}
+
+var carColumns = []carColumn{
+	{
+		"Car",
+		"STRING",
+		func(c *Car, val string) error {
+			c.Car = val
+			return nil
+		},
+		func(c *Car) (string, error) {
+			return c.Car, nil
+		},
+	},
+	{
+		"MPG",
+		"DOUBLE",
+		func(c *Car, val string) (err error) {
+			c.MPG, err = strconv.ParseFloat(val, 64)
+			return
+		},
+		func(c *Car) (string, error) {
+			return fmt.Sprintf("%f", c.MPG), nil
+		},
+	},
+	{
+		"Cylinders",
+		"INT",
+		func(c *Car, val string) (err error) {
+			c.Cylinders, err = strconv.ParseInt(val, 10, 64)
+			return
+		},
+		func(c *Car) (string, error) {
+			return fmt.Sprintf("%d", c.Cylinders), nil
+		},
+	},
+	// remaining columns ...
+}
+```
+
+This is of course a bit verbose, but on the other hand it solves our problem. Reordering our columns is now a single cut & paste operation, and adding or removing a column has gotten a lot easier as well. The approach even gives us a convenient way to put the `Type` information found in the second line of the CSV file, so we can easily duplicate this quirk when writing our own CSV files later on.
+
+Of course we still need to update our `UnmarshalRecord` and `MarshalRecord` from earlier. The good news is that we're unlikely to ever have to modify this code again:
+
+```go
+func (c *Car) UnmarshalRecord(record []string) error {
+	if got, want := len(record), len(carColumns); got != want {
+		return fmt.Errorf("bad number of columns: got=%d want=%d", got, want)
+	}
+	for i, col := range carColumns {
+		if err := col.UnmarshalValue(c, record[i]); err != nil {
+			return fmt.Errorf("column=%q: %w", col.Name, err)
+		}
+	}
+	return nil
+}
+
+func (c *Car) MarshalRecord() ([]string, error) {
+	record := make([]string, len(carColumns))
+	for i, col := range carColumns {
+		val, err := col.MarshalValue(c)
+		if err != nil {
+			return nil, err
+		}
+		record[i] = val
+	}
+	return record, nil
+}
+```
+
+Now it's time to implement the CSV decoding itself. My boilerplate implements this via the `ReadCarsCSV` function that can be seen below.
+
+```go
+const carComma = ';'
+
+func ReadCarsCSV(r io.Reader) ([]*Car, error) {
+	cr := csv.NewReader(r)
+	cr.Comma = carComma
+
+	records, err := cr.ReadAll()
+	if err != nil {
+		return nil, err
+	}
+
+	var cars []*Car
```

Comment:
Premature optimization.

```suggestion
	cars := make([]*Car, 0, len(records))
```

Reply:
Hm, ok. I usually try to avoid this in my own code because it creates more entanglement between things that is often not worth it. But OTOH I mentioned the performance benefits of the boilerplate approach, so I guess hinting at these optimizations isn't a bad idea here. Thx!